src/main/java/picard/illumina/SampleInputParameters.java

```diff
@@ -0,0 +1,119 @@
+import picard.util.IlluminaUtil;
+import picard.util.TabbedTextFileWithHeaderParser;
+
+import java.io.File;
+import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * This class encloses sample input parameters used for any ExtractBarcodesProgram (ExtractIlluminaBarcodes,
+ * IlluminaBasecallsToFastq, IlluminaBasecallsToSam). This is used to parse tab delimited sample information including
+ * barcodes, library names and output prefixes.
+ */
+public class SampleInputParameters {
+    /**
+     * Column header for the first barcode sequence (preferred).
+     */
+
+    public static final Set<String> BARCODE_PREFIXES = new HashSet<>(
+            Arrays.asList(SampleInputParameters.BARCODE_SEQUENCE_COLUMN, SampleInputParameters.BARCODE_COLUMN)
+    );
+
+    public static final String BARCODE_COLUMN = "barcode";
+    public static final String BARCODE_SEQUENCE_COLUMN = "barcode_sequence";
+
+    /**
+     * Column header for the barcode name.
+     */
+    public static final String BARCODE_NAME_COLUMN = "barcode_name";
+    /**
+     * Column header for the library name.
+     */
+    public static final String LIBRARY_NAME_COLUMN = "library_name";
+
+
+    /**
+     * Parses any one of the following types of files:
+     *
+     * ExtractIlluminaBarcodes      BARCODE_FILE
+     * IlluminaBasecallsToFastq     MULTIPLEX_PARAMS
+     * IlluminaBasecallsToSam       LIBRARY_PARAMS
+     *
+     * This will validate to file format as well as populate a Map of barcodes to metrics.
+     *
+     * @param inputFile         The input file that is being parsed
+     * @param readStructure     The read structure for the reads of the run
+     */
+    protected static Tuple<Map<String, BarcodeMetric>, List<String>> parseInputFile(final File inputFile,
+                                          final ReadStructure readStructure) {
+        List<String> messages = new ArrayList<>();
+        Map<String, BarcodeMetric> barcodeToMetrics = new LinkedHashMap<>();
+        try (final TabbedTextFileWithHeaderParser barcodesParser = new TabbedTextFileWithHeaderParser(inputFile)) {
+            List<String> validBarcodeColumns = barcodesParser.columnLabels().stream().filter(name -> {
+                boolean isValidPrefix = false;
+                for (String columnPrefix : BARCODE_PREFIXES) {
+                    isValidPrefix |= name.toUpperCase().startsWith(columnPrefix.toUpperCase());
+                }
+                return isValidPrefix;
+            }).collect(Collectors.toList());
+
+            if (readStructure.sampleBarcodes.length() != validBarcodeColumns.size()) {
+                messages.add("Expected " + readStructure.sampleBarcodes.length() + " valid barcode columns, but only found " +
+                        String.join(",", validBarcodeColumns));
+            }
+
+            Matcher matcher = Pattern.compile("^(.*)_\\d").matcher(validBarcodeColumns.get(0));
```

Comment:
Do you care which one you get?  Might it matter if the user has put `BARCODE_2` before `BARCODE`?  If so might be good to sort `validBarcodeColumns` first by length and then by `String.compare()`?